src/ke.cpp

```diff
@@ -330,21 +352,22 @@ KeQueryActiveProcessorCountEx(_In_ USHORT group_number) { return KeQueryMaximumP
 KAFFINITY
 KeSetSystemAffinityThreadEx(KAFFINITY affinity)
 {
+    GROUP_AFFINITY new_affinity = {0};
+    new_affinity.Mask = affinity;
+    new_affinity.Group = _usersim_group_affinity_cache.Group;
     GROUP_AFFINITY old_affinity;
-    usersim_get_current_thread_group_affinity(&old_affinity);
-    // Reject affinities that are not valid for the current group.
-    // Assume that the affinity mask is contiguous.
-    KAFFINITY valid_affinity_mask = (1 << KeQueryMaximumProcessorCountEx(old_affinity.Group)) - 1;
-    if ((affinity & valid_affinity_mask) == 0) {
-        return 0;
-    }
-    _usersim_thread_affinity.Group = old_affinity.Group;
-    _usersim_thread_affinity.Mask = affinity;
-    if (KeGetCurrentIrql() < DISPATCH_LEVEL && SetThreadAffinityMask(GetCurrentThread(), affinity) == 0) {
-        _usersim_thread_affinity = old_affinity;
+
+    if (KeGetCurrentIrql() >= DISPATCH_LEVEL) {
+        _usersim_group_affinity_cache = new_affinity;
         return 0;
+    } else {
+        bool result = usersim_set_current_thread_affinity(&new_affinity, &old_affinity);
+        if (result) {
+            return (KAFFINITY)old_affinity.Mask;

```

Comment:
I'm not sure why this cast is required - is `old_affinity.Mask` not already a `KAFFINITY`?